br/pkg/lightning/restore/check_info.go

```diff
@@ -658,86 +660,260 @@ func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *mydump.MDTab
-				}
+	// compare column names and make sure
+	// 1. TiDB table info has data file's all columns(besides ignore columns)
+	// 2. Those columns not introduced in data file always have a default value.
+	colMap := make(map[string]struct{})
+	for col := range igCols {
+		colMap[col] = struct{}{}
+	}
+	for _, col := range core.Columns {
+		if _, ok := colMap[col.Name.L]; ok {
+			// tidb's column is ignored
+			// we need ensure this column has the default value.
+			if _, hasDefault := defaultCols[col.Name.L]; !hasDefault {
+				msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s`'s column %s cannot be ignored,"+
+					"because it doesn't have a default value, please set tables.ignoreColumns properly",
+					tableInfo.DB, tableInfo.Name, col.Name.L))
 			}
 		} else {
-			// compare column names and make sure
-			// 1. TiDB table info has data file's all columns(besides ignore columns)
-			// 2. Those columns not introduced in data file always have a default value.
-			colMap := make(map[string]struct{})
-			for col := range igCols {
-				colMap[col] = struct{}{}
+			colMap[col.Name.L] = struct{}{}
+		}
+	}
+	// tidb_rowid can be ignored in check
+	colMap[model.ExtraHandleName.String()] = struct{}{}
+	for _, col := range colsFromDataFile {
+		if _, ok := colMap[col]; !ok {
+			checkMsg := "please check table schema"
+			if dataFileMeta.Type == mydump.SourceTypeCSV && rc.cfg.Mydumper.CSV.Header {
+				checkMsg += " and csv file header"
 			}
-			for _, col := range core.Columns {
-				if _, ok := colMap[col.Name.L]; ok {
-					// tidb's column is ignored
-					// we need ensure this column has the default value.
-					if _, hasDefault := defaultCols[col.Name.L]; !hasDefault {
-						msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s`'s column %s cannot be ignored,"+
-							"because it doesn't hava a default value, please set tables.ignoreColumns properly",
-							tableInfo.DB, tableInfo.Name, col.Name.L))
-					}
-				} else {
-					colMap[col.Name.L] = struct{}{}
-				}
+			msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have column %s, "+
+				"%s or use tables.ignoreColumns to ignore %s",
+				tableInfo.DB, tableInfo.Name, col, checkMsg, col))
+		} else {
+			// remove column for next iteration
+			delete(colMap, col)
+		}
+	}
+	// if theses rest columns don't have a default value.
+	for col := range colMap {
+		if _, ok := defaultCols[col]; ok {
+			continue
+		}
+		msgs = append(msgs, fmt.Sprintf("TiDB schema `%s`.`%s` doesn't have the default value for %s"+
+			"please give a default value for %s or choose another column to ignore or add this column in data file",
+			tableInfo.DB, tableInfo.Name, col, col))
+	}
+	return msgs, nil
+}
+
+// checkCSVHeader try to check whether the csv header config is consistent with the source csv files
+func (rc *Controller) checkCSVHeader(ctx context.Context, dbMetas []*mydump.MDDatabaseMeta) error {
```

Comment:
i suggest adding a comment to outline the algorithm used in the check.

```
1. pick one table with two CSV files and a unique/primary key
2. read the first row of those two CSV files
3. checks if the content of those first rows are compatible with the table schema, and whether the two rows are identical, to determine if the first rows are a header rows.
```

Reply:
done